ansible_base/resource_registry/registry.py

```diff
@@ -0,0 +1,102 @@
+
+
+class ServiceAPIConfig:
+    service_type = None
+
+    role_definitions_viewset = None
+    user_role_assignment_viewset = None
+    team_role_assignment_viewset = None
+    permission_viewset = None
+
+    permission_class = permissions.IsAuthenticated
+
+    def authorize_user(username, password=None, ansible_id=None):
+        raise NotImplementedError
+
+
+class ResourceInspector(BaseSchemaGenerator):
+    def __init__(self, urlpatterns=None):
+        super().__init__(patterns=urlpatterns)
+
+        self._initialise_endpoints()
+        self.model_map = {}
+
+        for path, method, callback in self.endpoints:
+            view = self.create_view(callback, method)
+
+            if hasattr(view, "get_serializer_class"):
+                serializer_class = view.get_serializer_class()
+                if hasattr(serializer_class, "Meta"):
+                    label = serializer_class.Meta.model._meta.label
+                    if label not in self.model_map:
+                        self.model_map[label] = {}
+
+                    if hasattr(view, "action"):
+                        action = view.action
+                        if action == "partial_update":
+                            action = "update"
+                        if action not in self.model_map[label]:
+                            self.model_map[label][action] = []
+                        self.model_map[label][action].append((method, path))
+
+
+class ResourceRegistry:
+    registry = {}
+
+    def __init__(self, service_api_config: ServiceAPIConfig = None):
+        self.resource_inspector = ResourceInspector()
+
+        self._validate_api_config(service_api_config)
+        self.api_config = service_api_config
+
+    def _validate_api_config(self, config):
+        """
+        Needs to validate that:
+            - Viewsets have the correct serializer, pagination and filter classes
+            - Service type is set to one of awx, galaxy, eda or gateway
+        """
+        assert config.service_type in ["aap", "awx", "galaxy", "eda"]
+
+    def register(self, model, shared_resource: SharedResource = None, parent_resources: List[ParentResource] = None, name_field: str = None):
+        model_label = model._meta.label
+        managed_serializer = None
+        externally_managed = False
+        if name_field is None:
+            name_field = "name"
+
+        if shared_resource:
+            managed_serializer = shared_resource.serializer
+            externally_managed = not shared_resource.is_provider
+
+        parent_map = {}
+        if parent_resources:
+            for parent in parent_resources:
+                parent_map[parent.model._meta.label] = parent
+
+        self.registry[model_label] = {
+            "model": model,
+            "externally_managed": externally_managed,
+            "managed_serializer": managed_serializer,
+            "parent_resources": parent_map,
```

Comment:
Is `parent_resources` for a feature that doesn't yet exist?

Reply:
yeah, that will be used for RBAC. I might get rid of it for the time being, depending on what @AlanCoding thinks.